### ISSUE TITLE
Error handling for Set Value

### ISSUE THREAD

**Issue by johndanenRV** [rossvideo/Catena#258]
We need to decide how the fromProto method should handle request from the client that are partially invalid. i.e. if the client sends a set value request for a struct and one of the struct_value fields given is a field that the client is not authorized to write to, or it does not exist within the param, then how should the error be handled.

**Option 1: throw an error**
If the request includes an invalid field, then throw an error and cancel the request with bad status. In order to avoid misleading error messages, the bad status response from the service would include json pointer to whatever field caused the error.

**Option 2: Ignore invalid fields**
If the request includes an invalid field, then fromProto will simply ignore that field and continue to write any valid fields from the incoming value. The client would then see in the update pushed by the connect rpc that the invalid field did not change.

**Comment by johndanenRV** [comment]
Notes from today's meeting:
- In the case of a partially valid request, the valid parts of the request will still be processed. The invalid parts of the request will be skipped, and error info should be logged and passed to the client. The call should end with status::ok.
- For functions that need to be able to return both a value and an error status, we will define a result info struct so that both can be returned together. 
  - The implementation of this should include documentation to clarify that this makes use of [C++ return value optimization](https://stackoverflow.com/questions/12953127/what-are-copy-elision-and-return-value-optimization/12953129#12953129) to avoid making unnecessary copies of the return value and the error status
  - for cases where the error is fatal and there is no param to return, we can use [std::optional](https://en.cppreference.com/w/cpp/utility/optional) to return std::nullopt instead of returning null pointers
- Important implementation detail to avoid information leaking to attackers: 
  - trying to write to a param that a client DOES have read access for should return an unauthorized error status. 
  - trying to read/write to a param that a client DOES NOT have read access for should return a does not exist error status.

**Comment by johndanenRV** [comment]
I've been reading the [gRPC Error Handling](https://grpc.io/docs/guides/error/) documentation. Currently we are using the standard error model provided gRPC but the documentation also talks about a richer error model provided by Protobuf. I am wondering if this model would be better suited for us because, as the documentation states, "This model enables servers to return and clients to consume additional error details expressed as one or more protobuf messages." This would be useful because set value request can include an arbitrary number of invalid sub-params so it would be better to send the error details as multiple protobuf messages.

There are some drawbacks also explained in the documentation, but this would be worth considering. With the generic gRPC error handling it will be difficult to provide detailed error info to clients.

**Comment by mejohnnaylor** [comment]
I think we need to keep it simple and move on to bigger issues.

BTW the guidelines in status.h advise these return codes

- UNAUTHENTICATED if access is unauthorized, but the id of the user isn't known
- PERMISSION_DENIED if user id is known, but access is not authorized

I'm not particularly concerned about returning live, detailed error info to clients because, as stated, we have bigger issues. However, we should be logging these events so that a log management system can get at them for after the fact failure analysis.

In summary - if any part of the SetValue request succeeds, we return status::OK and stream the values of updated parameters to which the attached client has access, and log the unauthorized accesses locally.
